File: leetcode-cn/3.无重复字符的最长子串.py
#
# @lc app=leetcode.cn id=3 lang=python3
#
# [3] 无重复字符的最长子串
#

# @lc code=start
from collections import defaultdict


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        right = 0
        window = defaultdict(int)
        ans = 0

        while right < len(s):
            c = s[right]
            right += 1

            window[c] += 1

            while window[c] > 1:
                d = s[left]
                left += 1

                window[d] -= 1

            ans = max(ans, right - left)

        return ans


# @lc code=end

File: leetcode-cn/C++/3.无重复字符的最长子串.cpp
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
#include <string>
#include <unordered_set>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s)
    {
        int ans = 0;

        int left = 0;
        int right = 0;

        unordered_set<char> window;

        while (right < s.size()) {
            while (window.count(s[right])) {
                window.erase(s[left]);
                left++;
            }
            window.insert(s[right]);
            right++;
            ans = max(ans, right - left);
        }

        return ans;
    }
};

// @lc code=end
